Circle CI upgrade to 2.0

Denis-Gavrielov opened the pull request:
Upgrade from Circle CI 1.0 to version 2.0

Isolate the dependencies for the e2e tests into one Docker image for Python tests and one Docker image for JavaScript tests.
Rewrite the JavaScript end to end test to use chrome headless with Puppeteer and jest instead of the depreciated phantomjs library.

ccadar commented:
@Denis-Gavrielov I see the CircleCI tests still fail with these patches?

Denis-Gavrielov commented:
@ccadar yes, and this PR will need to be overworked a little bit before it is actually merged.

These tests can sometimes work and sometimes timeout (in both the local dev and circle CI environment), so I am assuming that there is something off in my JS code in terms of the stability of the tests (as I am fairly unfamiliar with JS). I believe that they are being run in parallel or a random order and that causes issues.
I am hoping to receive some feedback from @andronat later this week to make these tests stable so they stop failing.

Denis-Gavrielov commented:
@ccadar @andronat the tests should be stable in dev and circle CI now. 

Feel free to suggest changes, but otherwise it should be good to be merged in.

ccadar commented:
@Denis-Gavrielov Really great to see the tests working again, great job!  
I noticed though, that some tests, such as `get_sign`, were removed.  Why is that?  Or did you re-add them in a different way and I missed that?

Denis-Gavrielov commented:
@ccadar Thank you! 
I have split the Python and JavaScript tests into two separate containers now. They both run on Circle CI, first the Python tests (including Flake8 and testing if the web server is up), then the JavaScript tests. You can individually inspect and see that they all pass by clicking on the last two commands that ran in Circle CI.
https://circleci.com/gh/Denis-Gavrielov/klee-web/255?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
Here you can inspect the Python tests command: 
`sudo docker run --rm -it --network host -v /var/run/docker.sock:/var/run/docker.sock -v /tmp/:/tmp/ -v /titb/:/titb/ e2e_test`
and JavaScript tests command:
`sudo docker run --rm -it --network host -v /titb/src/klee_web/tests/js_tests/test_files/:/titb/src/klee_web/tests/js_tests/test_files/ e2e_test_js`

andronat commented on src/klee_web/tests/jest-puppeteer.config.js:
```diff
@@ -0,0 +1,8 @@
+      '--disable-setuid-sandbox',
+    ]
+  }
+}
```
Add EOF

andronat commented on provisioning/roles/webdriver/tasks/main.yml:
```diff
@@ -18,3 +18,9 @@
   tags: e2e
+
+- name: build js e2e tests image
+  command: docker build --file ./tests/Dockerfile_js -t e2e_test_js .
```
Is there a reason you use `command` instead of ansible's native docker plugin? Read [here](https://docs.ansible.com/ansible/latest/modules/docker_image_module.html).

Denis-Gavrielov replied:
Sometimes the native docker plugin would run into issues. I tried again and this time it seems to work well.

andronat commented on src/klee_web/tests/Dockerfile_js:
```diff
@@ -0,0 +1,41 @@
+# Run everything after as non-privileged user.
+USER pptruser
+
+CMD npm test
```
Why not [this](https://hub.docker.com/r/buildkite/puppeteer/dockerfile)?

Denis-Gavrielov replied:
I have tried multiple options running Puppeteer inside a container including this one. They all failed except the one I am using right now. 
I took it from Puppeteer's github page under section **Running Puppeteer in Docker**
https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md#running-puppeteer-in-docker

andronat commented on src/klee_web/worker/tests/fixtures/invalid_syntax/expected.stdout:
```diff
@@ -1,6 +1,2 @@
-1 warning and 1 error generated.
\ No newline at end of file
+((.*?|\n)+)warning: missing terminating '"' character \[-Winvalid-pp-token\]
+((.*?|\n)+)error: expected expression((.*?|\n)+)expected '}((.*?|\n)+)to match this '{'
```
Add EOF

Denis-Gavrielov replied:
These expected outputs for tests seem to fail when an EOF is added, as they are being compared to the actual output produced by the website.
I have noticed that in the previous commits before my changes there were no EOF. In that case, would it make sense to leave it as it is (also applying to the other cases where there is no EOF in the `expected.stdout` files)?

andronat commented on src/klee_web/worker/tests/fixtures/simple_unicode/expected.stdout:
```diff
@@ -1,6 +1,5 @@
 KLEE: done: completed paths = 1
-KLEE: done: generated tests = 1
\ No newline at end of file
+KLEE: done: generated tests = 1$
```
Add EOF

andronat commented on src/klee_web/worker/tests/fixtures/symargs/expected.stdout:
```diff
@@ -3,5 +3,5 @@
-KLEE: done: generated tests = 2
\ No newline at end of file
+KLEE: done: completed paths = 3
+KLEE: done: generated tests = 3
```
Add EOF

andronat commented on provisioning/roles/webdriver/tasks/main.yml:
```diff
@@ -2,25 +2,25 @@
 
 - name: Build E2E_test container
-  command: docker build --file ./src/klee_web/tests/Dockerfile -t e2e_test .
+  command: docker build --file ./src/klee_web/tests/python_tests/Dockerfile -t e2e_test .
```
Use ansible's docker plugin here.

andronat commented on src/klee_web/tests/python_tests/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+# WORKDIR /titb/src/klee_web/tests/python_tests
+
+# ## be able to launch docker
+RUN curl -sSL https://get.docker.com/ | sh
```
Why do we need this?

Denis-Gavrielov replied:
Please above for an explanation on why we need to run Docker commands for the Python unit tests.

andronat commented on src/klee_web/tests/python_tests/klee-web-environment.sh:
```diff
@@ -0,0 +1,12 @@
+export DB_HOST=localhost
+export DB_PORT=5432
+export DB_NAME=klee_web
+export DEVELOPMENT=1
```
I left a comment before, on how you should do this.

Denis-Gavrielov replied:
Please see above for a comment on how this is actually resolved dynamically as you suggested.

andronat commented on src/klee_web/tests/js_tests/test_files/e2e.test.js:
```diff
@@ -5,7 +5,7 @@ jest.setTimeout(15000);
   beforeAll(async () => {
 
-    await page.goto('http://192.168.33.10/');
+    await page.goto('http://localhost/');
```
make these dynamic

Denis-Gavrielov replied:
These are fixed at localhost for this commit. I am not sure what you mean exactly.

In the dockerized PR these will be environmental variables that are pushed into the container as you have suggested.

andronat commented on src/klee_web/tests/js_tests/test_files/e2e.test.js:
```diff
@@ -1,14 +1,15 @@
     await page.click('[ng-repeat="file in files"]');  // select first file ... use page.select instead?
     await page.click('#run-klee-btn');
-    await page.waitFor('#result-output');
+    await new Promise(resolve => setTimeout(resolve, 15000));
```
Let's not use timeouts. We did this together just to debug what is the problem. We should be using proper waits like [this](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagewaitforselectorselector-options).

Denis-Gavrielov replied:
Done, thanks for the link!

andronat commented on provisioning/roles/web/tasks/deploy.yml:
```diff
@@ -54,6 +54,12 @@
+         && . /etc/profile.d/klee-web-environment.sh 
+         && cd {{ application_dir }}
+         && python manage.py makemigrations
+
```
We need to restructure the commits in general. This commit can actually go first. You can even take it out completely and make it a separate PR so we can merge it fast.

andronat commented on provisioning/vagrant.yml:
```diff
@@ -22,4 +22,6 @@
-    - webdriver
+    - role: e2e-tests
+      when: ci|bool
+
```
Remove this empty line. Don't remove the EOF.

Denis-Gavrielov commented on .circleci/config.yml:
```diff
@@ -0,0 +1,18 @@
+    steps:
+      - checkout  
+      - run: sudo apt-get install software-properties-common
+      - run: sudo apt-get update -qq
```
I found that both works. I have switched the position of both lines.

Denis-Gavrielov commented on provisioning/roles/web/tasks/deploy.yml:
```diff
@@ -27,20 +27,33 @@
 
+- name: remove .gitconfig in ci
+  file: 
+    path: ~/.gitconfig
```
This was related to an error that occurred within circle ci. 
The .gitconfig has the following content:
```
[url "ssh://git@github.com"]
	insteadOf = https://github.com
[gc]
	auto = 0
```

which then runs into problems e.g. at the provisioning step ```web : Get bower dependencies``` when GitHub needs to be reached.

The error log is:
```
TASK [web : Get bower dependencies] ********************************************
task path: /home/circleci/project/provisioning/roles/web/tasks/deploy.yml:36
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952 `" && echo ansible-tmp-1565086069.07-124584845988952="` echo /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952 `" ) && sleep 0'
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/commands/command.py
<localhost> PUT /home/circleci/.ansible/tmp/ansible-local-83106v0pav/tmprMNWSB TO /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952/AnsiballZ_command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952/ /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1565086069.07-124584845988952/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": true, 
    "cmd": [
        "bower", 
        "install", 
        "--config.interactive=false", 
        "--allow-root"
    ], 
    "delta": "0:00:01.810924", 
    "end": "2019-08-06 10:07:50.990256", 
    "invocation": {
        "module_args": {
            "_raw_params": "bower install --config.interactive=false --allow-root", 
            "_uses_shell": false, 
            "argv": null, 
            "chdir": "/titb", 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "stdin": null, 
            "stdin_add_newline": true, 
            "strip_empty_ends": true, 
            "warn": true
        }
    }, 
    "msg": "non-zero return code", 
    "rc": 1, 
    "start": "2019-08-06 10:07:49.179332", 
    "stderr": "bower nanobar#~0.2.0                          ECMDERR Failed to execute \"git ls-remote --tags --heads https://github.com/jacoborus/nanobar.git\", exit code of #128 Warning: Permanently added 'github.com,192.30.253.113' (RSA) to the list of known hosts. Permission denied (publickey). fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.\n\nAdditional error details:\nWarning: Permanently added 'github.com,192.30.253.113' (RSA) to the list of known hosts.\r\nPermission denied (publickey).\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.", 
    "stderr_lines": [
        "bower nanobar#~0.2.0                          ECMDERR Failed to execute \"git ls-remote --tags --heads https://github.com/jacoborus/nanobar.git\", exit code of #128 Warning: Permanently added 'github.com,192.30.253.113' (RSA) to the list of known hosts. Permission denied (publickey). fatal: Could not read from remote repository.  Please make sure you have the correct access rights and the repository exists.", 
        "", 
        "Additional error details:", 
        "Warning: Permanently added 'github.com,192.30.253.113' (RSA) to the list of known hosts.", 
        "Permission denied (publickey).", 
        "fatal: Could not read from remote repository.", 
        "", 
        "Please make sure you have the correct access rights", 
        "and the repository exists."
    ], 
    "stdout": "bower bootstrap-social#~5.0.0       not-cached https://github.com/lipis/bootstrap-social.git#~5.0.0\nbower bootstrap-social#~5.0.0          resolve https://github.com/lipis/bootstrap-social.git#~5.0.0\nbower animate.css#~3.2.0            not-cached https://github.com/daneden/animate.css.git#~3.2.0\nbower animate.css#~3.2.0               resolve https://github.com/daneden/animate.css.git#~3.2.0\nbower font-awesome#*                not-cached https://github.com/FortAwesome/Font-Awesome.git#*\nbower font-awesome#*                   resolve https://github.com/FortAwesome/Font-Awesome.git#*\nbower angular-resource#~1.3.8       not-cached https://github.com/angular/bower-angular-resource.git#~1.3.8\nbower angular-resource#~1.3.8          resolve https://github.com/angular/bower-angular-resource.git#~1.3.8\nbower bootstrap-sass-official#~3.3.1       not-cached https://github.com/twbs/bootstrap-sass.git#~3.3.1\nbower bootstrap-sass-official#~3.3.1          resolve https://github.com/twbs/bootstrap-sass.git#~3.3.1\nbower elegant-icons#*                      not-cached https://github.com/josephnle/elegant-icons.git#*\nbower elegant-icons#*                         resolve https://github.com/josephnle/elegant-icons.git#*\nbower angular-ui-codemirror#~0.2.1         not-cached https://github.com/angular-ui/ui-codemirror.git#~0.2.1\nbower angular-ui-codemirror#~0.2.1            resolve https://github.com/angular-ui/ui-codemirror.git#~0.2.1\nbower angular-animate#~1.3.8               not-cached https://github.com/angular/bower-angular-animate.git#~1.3.8\nbower angular-animate#~1.3.8                  resolve https://github.com/angular/bower-angular-animate.git#~1.3.8\nbower angular#~1.3.8                       not-cached https://github.com/angular/bower-angular.git#~1.3.8\nbower angular#~1.3.8                          resolve https://github.com/angular/bower-angular.git#~1.3.8\nbower jquery#~2.1.3                        not-cached https://github.com/jquery/jquery-dist.git#~2.1.3\nbower jquery#~2.1.3                           resolve https://github.com/jquery/jquery-dist.git#~2.1.3\nbower angular-cookies#~1.3.8               not-cached https://github.com/angular/bower-angular-cookies.git#~1.3.8\nbower angular-cookies#~1.3.8                  resolve https://github.com/angular/bower-angular-cookies.git#~1.3.8\nbower angular-ui-slider#~0.0.2             not-cached https://github.com/angular-ui/ui-slider.git#~0.0.2\nbower angular-ui-slider#~0.0.2                resolve https://github.com/angular-ui/ui-slider.git#~0.0.2\nbower underscore#~1.7.0                    not-cached https://github.com/jashkenas/underscore.git#~1.7.0\nbower underscore#~1.7.0                       resolve https://github.com/jashkenas/underscore.git#~1.7.0\nbower nanobar#~0.2.0                       not-cached https://github.com/jacoborus/nanobar.git#~0.2.0\nbower nanobar#~0.2.0                          resolve https://github.com/jacoborus/nanobar.git#~0.2.0\nbower angular-file-upload#~1.1.5           not-cached https://github.com/nervgh/angular-file-upload.git#~1.1.5\nbower angular-file-upload#~1.1.5              resolve https://github.com/nervgh/angular-file-upload.git#~1.1.5\nbower angular-bootstrap#~0.12.0            not-cached https://github.com/angular-ui/bootstrap-bower.git#~0.12.0\nbower angular-bootstrap#~0.12.0               resolve https://github.com/angular-ui/bootstrap-bower.git#~0.12.0", 
    "stdout_lines": [
        "bower bootstrap-social#~5.0.0       not-cached https://github.com/lipis/bootstrap-social.git#~5.0.0", 
        "bower bootstrap-social#~5.0.0          resolve https://github.com/lipis/bootstrap-social.git#~5.0.0", 
        "bower animate.css#~3.2.0            not-cached https://github.com/daneden/animate.css.git#~3.2.0", 
        "bower animate.css#~3.2.0               resolve https://github.com/daneden/animate.css.git#~3.2.0", 
        "bower font-awesome#*                not-cached https://github.com/FortAwesome/Font-Awesome.git#*", 
        "bower font-awesome#*                   resolve https://github.com/FortAwesome/Font-Awesome.git#*", 
        "bower angular-resource#~1.3.8       not-cached https://github.com/angular/bower-angular-resource.git#~1.3.8", 
        "bower angular-resource#~1.3.8          resolve https://github.com/angular/bower-angular-resource.git#~1.3.8", 
        "bower bootstrap-sass-official#~3.3.1       not-cached https://github.com/twbs/bootstrap-sass.git#~3.3.1", 
        "bower bootstrap-sass-official#~3.3.1          resolve https://github.com/twbs/bootstrap-sass.git#~3.3.1", 
        "bower elegant-icons#*                      not-cached https://github.com/josephnle/elegant-icons.git#*", 
        "bower elegant-icons#*                         resolve https://github.com/josephnle/elegant-icons.git#*", 
        "bower angular-ui-codemirror#~0.2.1         not-cached https://github.com/angular-ui/ui-codemirror.git#~0.2.1", 
        "bower angular-ui-codemirror#~0.2.1            resolve https://github.com/angular-ui/ui-codemirror.git#~0.2.1", 
        "bower angular-animate#~1.3.8               not-cached https://github.com/angular/bower-angular-animate.git#~1.3.8", 
        "bower angular-animate#~1.3.8                  resolve https://github.com/angular/bower-angular-animate.git#~1.3.8", 
        "bower angular#~1.3.8                       not-cached https://github.com/angular/bower-angular.git#~1.3.8", 
        "bower angular#~1.3.8                          resolve https://github.com/angular/bower-angular.git#~1.3.8", 
        "bower jquery#~2.1.3                        not-cached https://github.com/jquery/jquery-dist.git#~2.1.3", 

        "bower jquery#~2.1.3                           resolve https://github.com/jquery/jquery-dist.git#~2.1.3", 
        "bower angular-cookies#~1.3.8               not-cached https://github.com/angular/bower-angular-cookies.git#~1.3.8", 
        "bower angular-cookies#~1.3.8                  resolve https://github.com/angular/bower-angular-cookies.git#~1.3.8", 
        "bower angular-ui-slider#~0.0.2             not-cached https://github.com/angular-ui/ui-slider.git#~0.0.2", 
        "bower angular-ui-slider#~0.0.2                resolve https://github.com/angular-ui/ui-slider.git#~0.0.2", 
        "bower underscore#~1.7.0                    not-cached https://github.com/jashkenas/underscore.git#~1.7.0", 
        "bower underscore#~1.7.0                       resolve https://github.com/jashkenas/underscore.git#~1.7.0", 
        "bower nanobar#~0.2.0                       not-cached https://github.com/jacoborus/nanobar.git#~0.2.0", 
        "bower nanobar#~0.2.0                          resolve https://github.com/jacoborus/nanobar.git#~0.2.0", 
        "bower angular-file-upload#~1.1.5           not-cached https://github.com/nervgh/angular-file-upload.git#~1.1.5", 
        "bower angular-file-upload#~1.1.5              resolve https://github.com/nervgh/angular-file-upload.git#~1.1.5", 
        "bower angular-bootstrap#~0.12.0            not-cached https://github.com/angular-ui/bootstrap-bower.git#~0.12.0", 
        "bower angular-bootstrap#~0.12.0               resolve https://github.com/angular-ui/bootstrap-bower.git#~0.12.0"
    ]
}```

Denis-Gavrielov replied:
The .gitconfig file is placed there not by us, but by circle CI.

Denis-Gavrielov commented on provisioning/roles/init/tasks/main.yml:
```diff
@@ -5,10 +5,18 @@
+- name: Update and install language-pack-en-base
+  apt:
+    name:
+      - language-pack-en-base
```
Circle CI would fail with this line. I have just tested it and it seems like we don't need it in dev. I am removing this installation completely.

Denis-Gavrielov commented on provisioning/vagrant.yml:
```diff
@@ -15,7 +15,8 @@
       pip_package: python-pip
-    - geerlingguy.docker
+    - role: geerlingguy.docker
+      when: not ci
```
The geerlingguy.docker role crashes in Circle CI; however, Docker is already installed in the environment so it is not needed to be installed.

Denis-Gavrielov commented on provisioning/roles/web/tasks/deploy.yml:
```diff
@@ -27,20 +27,33 @@
+    path: /etc/boto.cfg
+    state: absent
+  become: yes
+  when: ci|bool
```
The boto.cfg file crashes the build. 
I found a related post here: https://discuss.circleci.com/t/install-of-boto-now-broken-in-2-0-machine-build/18459

here is the error coming from circle CI:
```
TASK [web : Make Django migrations] ********************************************
task path: /home/circleci/project/provisioning/roles/web/tasks/deploy.yml:57
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373 `" && echo ansible-tmp-1565087260.25-181834194410373="` echo /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373 `" ) && sleep 0'
Using module file /usr/local/lib/python2.7/dist-packages/ansible/modules/commands/command.py
<localhost> PUT /home/circleci/.ansible/tmp/ansible-local-8333tJ3LUs/tmpBbGoqR TO /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373/AnsiballZ_command.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373/ /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373/AnsiballZ_command.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1565087260.25-181834194410373/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": true, 
    "cmd": ". /src/web/env/bin/activate && . /etc/profile.d/klee-web-environment.sh && cd /titb/src/klee_web && python manage.py makemigrations", 
    "delta": "0:00:01.368805", 
    "end": "2019-08-06 10:27:41.720834", 
    "invocation": {
        "module_args": {
            "_raw_params": ". /src/web/env/bin/activate && . /etc/profile.d/klee-web-environment.sh && cd /titb/src/klee_web && python manage.py makemigrations", 
            "_uses_shell": true, 
            "argv": null, 
            "chdir": null, 
            "creates": null, 
            "executable": null, 
            "removes": null, 
            "stdin": null, 
            "stdin_add_newline": true, 
            "strip_empty_ends": true, 
            "warn": true
        }
    }, 
    "msg": "non-zero return code", 
    "rc": 1, 
    "start": "2019-08-06 10:27:40.352029", 
    "stderr": "Traceback (most recent call last):\n  File \"manage.py\", line 10, in <module>\n    execute_from_command_line(sys.argv)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/__init__.py\", line 381, in execute_from_command_line\n    utility.execute()\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/__init__.py\", line 375, in execute\n    self.fetch_command(subcommand).run_from_argv(self.argv)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 323, in run_from_argv\n    self.execute(*args, **cmd_options)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 361, in execute\n    self.check()\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 390, in check\n    include_deployment_checks=include_deployment_checks,\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 377, in _run_checks\n    return checks.run_checks(**kwargs)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/registry.py\", line 72, in run_checks\n    new_errors = check(app_configs=app_configs)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/urls.py\", line 40, in check_url_namespaces_unique\n    all_namespaces = _load_all_namespaces(resolver)\n  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/urls.py\", line 57, in _load_all_namespaces\n    url_patterns = getattr(resolver, 'url_patterns', [])\n  File \"/src/web/env/lib/python3.5/site-packages/django/utils/functional.py\", line 80, in __get__\n    res = instance.__dict__[self.name] = self.func(instance)\n  File \"/src/web/env/lib/python3.5/site-packages/django/urls/resolvers.py\", line 579, in url_patterns\n    patterns = getattr(self.urlconf_module, \"urlpatterns\", self.urlconf_module)\n  File \"/src/web/env/lib/python3.5/site-packages/django/utils/functional.py\", line 80, in __get__\n    res = instance.__dict__[self.name] = self.func(instance)\n  File \"/src/web/env/lib/python3.5/site-packages/django/urls/resolvers.py\", line 572, in urlconf_module\n    return import_module(self.urlconf_name)\n  File \"/src/web/env/lib/python3.5/importlib/__init__.py\", line 126, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 986, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 969, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 958, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 673, in _load_unlocked\n  File \"<frozen importlib._bootstrap_external>\", line 665, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 222, in _call_with_frames_removed\n  File \"/home/circleci/project/src/klee_web/urls.py\", line 4, in <module>\n    import api\n  File \"/home/circleci/project/src/klee_web/api/__init__.py\", line 3, in <module>\n    from .views import ProjectViewSet, FileViewSet, JobViewSet\n  File \"/home/circleci/project/src/klee_web/api/views.py\", line 17, in <module>\n    from worker.worker import submit_code\n  File \"/home/circleci/project/src/klee_web/worker/worker.py\", line 10, in <module>\n    from .runner import WorkerRunner\n  File \"/home/circleci/project/src/klee_web/worker/runner.py\", line 19, in <module>\n    from .processor.upload import UploadProcessor\n  File \"/home/circleci/project/src/klee_web/worker/processor/upload.py\", line 6, in <module>\n    from worker.storage.s3_storage import S3Storage\n  File \"/home/circleci/project/src/klee_web/worker/storage/s3_storage.py\", line 3, in <module>\n    from boto.s3.key import Key\n  File \"/src/web/env/lib/python3.5/site-packages/boto/__init__.py\", line 1216, in <module>\n    boto.plugin.load_plugins(config)\n  File \"/src/web/env/lib/python3.5/site-packages/boto/plugin.py\", line 93, in load_plugins\n    _import_module(file)\n  File \"/src/web/env/lib/python3.5/site-packages/boto/plugin.py\", line 75, in _import_module\n    return imp.load_module(name, file, filename, data)\n  File \"/src/web/env/lib/python3.5/imp.py\", line 234, in load_module\n    return load_source(name, filename, file)\n  File \"/src/web/env/lib/python3.5/imp.py\", line 172, in load_source\n    module = _load(spec)\n  File \"/usr/lib/python3/dist-packages/google_compute_engine/boto/boto_config.py\", line 29, in <module>\n    from google_compute_engine import config_manager\nImportError: No module named 'google_compute_engine'", 
    "stderr_lines": [
        "Traceback (most recent call last):", 
        "  File \"manage.py\", line 10, in <module>", 
        "    execute_from_command_line(sys.argv)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/__init__.py\", line 381, in execute_from_command_line", 
        "    utility.execute()", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/__init__.py\", line 375, in execute", 
        "    self.fetch_command(subcommand).run_from_argv(self.argv)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 323, in run_from_argv", 
        "    self.execute(*args, **cmd_options)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 361, in execute", 
        "    self.check()", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 390, in check", 
        "    include_deployment_checks=include_deployment_checks,", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/management/base.py\", line 377, in _run_checks", 
        "    return checks.run_checks(**kwargs)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/registry.py\", line 72, in run_checks", 
        "    new_errors = check(app_configs=app_configs)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/urls.py\", line 40, in check_url_namespaces_unique", 
        "    all_namespaces = _load_all_namespaces(resolver)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/core/checks/urls.py\", line 57, in _load_all_namespaces", 
        "    url_patterns = getattr(resolver, 'url_patterns', [])", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/utils/functional.py\", line 80, in __get__", 
        "    res = instance.__dict__[self.name] = self.func(instance)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/urls/resolvers.py\", line 579, in url_patterns", 
        "    patterns = getattr(self.urlconf_module, \"urlpatterns\", self.urlconf_module)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/utils/functional.py\", line 80, in __get__", 
        "    res = instance.__dict__[self.name] = self.func(instance)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/django/urls/resolvers.py\", line 572, in urlconf_module", 
        "    return import_module(self.urlconf_name)", 
        "  File \"/src/web/env/lib/python3.5/importlib/__init__.py\", line 126, in import_module", 
        "    return _bootstrap._gcd_import(name[level:], package, level)", 
        "  File \"<frozen importlib._bootstrap>\", line 986, in _gcd_import", 
        "  File \"<frozen importlib._bootstrap>\", line 969, in _find_and_load", 
        "  File \"<frozen importlib._bootstrap>\", line 958, in _find_and_load_unlocked", 
        "  File \"<frozen importlib._bootstrap>\", line 673, in _load_unlocked", 
        "  File \"<frozen importlib._bootstrap_external>\", line 665, in exec_module", 
        "  File \"<frozen importlib._bootstrap>\", line 222, in _call_with_frames_removed", 
        "  File \"/home/circleci/project/src/klee_web/urls.py\", line 4, in <module>", 
        "    import api", 
        "  File \"/home/circleci/project/src/klee_web/api/__init__.py\", line 3, in <module>", 
        "    from .views import ProjectViewSet, FileViewSet, JobViewSet", 
        "  File \"/home/circleci/project/src/klee_web/api/views.py\", line 17, in <module>", 
        "    from worker.worker import submit_code", 
        "  File \"/home/circleci/project/src/klee_web/worker/worker.py\", line 10, in <module>", 
        "    from .runner import WorkerRunner", 
        "  File \"/home/circleci/project/src/klee_web/worker/runner.py\", line 19, in <module>", 
        "    from .processor.upload import UploadProcessor", 
        "  File \"/home/circleci/project/src/klee_web/worker/processor/upload.py\", line 6, in <module>", 
        "    from worker.storage.s3_storage import S3Storage", 
        "  File \"/home/circleci/project/src/klee_web/worker/storage/s3_storage.py\", line 3, in <module>", 
        "    from boto.s3.key import Key", 
        "  File \"/src/web/env/lib/python3.5/site-packages/boto/__init__.py\", line 1216, in <module>", 
        "    boto.plugin.load_plugins(config)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/boto/plugin.py\", line 93, in load_plugins", 
        "    _import_module(file)", 
        "  File \"/src/web/env/lib/python3.5/site-packages/boto/plugin.py\", line 75, in _import_module", 
        "    return imp.load_module(name, file, filename, data)", 
        "  File \"/src/web/env/lib/python3.5/imp.py\", line 234, in load_module", 
        "    return load_source(name, filename, file)", 
        "  File \"/src/web/env/lib/python3.5/imp.py\", line 172, in load_source", 
        "    module = _load(spec)", 
        "  File \"/usr/lib/python3/dist-packages/google_compute_engine/boto/boto_config.py\", line 29, in <module>", 
        "    from google_compute_engine import config_manager", 
        "ImportError: No module named 'google_compute_engine'"
    ], 
    "stdout": "", 
    "stdout_lines": []
}
```

Denis-Gavrielov commented on src/klee_web/tests/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+# ## be able to launch docker
+RUN curl -sSL https://get.docker.com/ | sh
+RUN apt install sudo
```
This must be an issue with the force pushes. I double checked and it is not installed more than once

Denis-Gavrielov commented on src/klee_web/klee-web-environment.sh:
```diff
@@ -0,0 +1,12 @@
+export DB_HOST=localhost
+export DB_PORT=5432
+export DB_NAME=klee_web
+export DEVELOPMENT=1
```
These variables are generated dynamically exactly by the files that you were referencing. 
What happens is that the very same files need to appear in the Docker container for the code to run.
So initially they are generated dynamically and then put into the machine (e.g. the vagrant machine in dev) and then these files are placed into the right folder within the vagrant machine so they can be passed into container.
E.g. the klee-web-environment.sh file is on the vagrant machine in /etc/profile.d/klee-web-environment.sh
However, this is outside the Docker build context, that is why I first have to copy it into the build context (in this case src/klee_web/) and then it can be passed into the container.

Copying the file into the right directory happens in provisioning/roles/e2e-tests/tasks/main.yml and passing in the file happens when the container is run.

Denis-Gavrielov replied:
Also, please note that I do not delete this file at the provisioning stage for a reason.

There is two ways to pass this files into the Docker container. One would be to use a COPY statement at build time, the other one is to pass it in via a volume. 
If I use a COPY statement and at some point we decide to push these images onto the Docker Hub, they will either have the development passwords in them (so won't work in the production environment), or we put production secrets in which means that our secrets are in a container on the Docker Hub.

That is why I chose to add this file into the container when it is run via a volume. This means that this file has to be present in the right directory on the vagrant machine while the container is running. 

I hope that makes sense.

Denis-Gavrielov commented on src/klee_web/tests/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY run_tests.sh /titb/
+
+# ## be able to launch docker
+RUN curl -sSL https://get.docker.com/ | sh
```
For the Python unit tests to work, the worker needs to launch the KLEE container.

What happens here is not that a Docker container is launched inside the testing container, but the  testing container is issuing a command to launch the KLEE container on the same level as the Docker container (so they are both managed by the Docker daemon inside the vagrant machine, but outside the testing container). 

Without the Docker cli none of the Python unit tests run.

Denis-Gavrielov commented on src/klee_web/tests/run_tests.sh:
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash -e
```
No reason. I have removed it now to avoid duplication.

Denis-Gavrielov commented on provisioning/roles/webdriver/tasks/main.yml:
```diff
@@ -1,10 +1,20 @@
-  pip: name=flake8
+- name: Copy klee-web-environment.sh file into klee_web directory
+  copy:
+    src: /etc/profile.d/klee-web-environment.sh
```
This is the file related to the discussion on passing in the environmental variables/secrets into the container via COPY or using the volume.

Denis-Gavrielov commented on src/klee_web/tests/.dockerignore:
```diff
@@ -0,0 +1,7 @@
+src/components
+src/sass
+src/static
+src/klee_web/node_modules
```
This file does not exist in the latest version of the PR.

Denis-Gavrielov commented on src/klee_web/tests/e2e_tests.js:
```diff
@@ -22,7 +22,7 @@ it('tests that input files are processed correctly', function(done1) {
       Promise.join(inputFuture, expectedOutputFuture, function(input, expected) {
           var n = new nightmare({load: 1000, webSecurity: false});
-          n.goto("http://localhost")
+          n.goto("http://192.168.33.10")
```
This is an outdated commit. Currently `localhost` is used and in the Dockerized version the IP address is passed in as an environmental variable.

Denis-Gavrielov commented on run_tests.sh:
```diff
@@ -25,7 +25,7 @@ DIR="$( cd "$( dirname "$0" )" && pwd )"
 _Run "Running Flake8 against Python Code" "flake8 --ignore=E722 --max-complexity 12 --exclude=migrations $DIR"
 
-_Run "Running Python Unit tests" "(source /src/worker/env/bin/activate && cd /titb/src/klee_web && /src/python_runner.sh python -m unittest discover -s worker/tests/ -p 'test_*.py')"
+_Run "Running Python Unit tests" "(cd /titb/src/klee_web && python -m unittest discover -s worker/tests/ -p 'test_*.py')"
```
I have tried multiple things to debug this. Yes, `_Run` does not let Circle CI report an error in general, but with the current setup it would only report a bug in Circle CI if the very last test in the run_test.sh file fails.
Therefore I have further split the tests of 

- Flake8
- Python Unit Tests
- webserver HTTP test

and they all run from their own shell script. I have tested each case excessively in Circle CI and it will flag out if either of those fails.

Denis-Gavrielov commented on provisioning/vars/development.yml:
```diff
@@ -1,6 +1,6 @@
 development: yes
 
-redis_host: localhost
+redis_host: 172.18.0.2  # replace this with kleeweb-redis once every component is dockerized
```
I have changed the command to this one:
`redis_host: "{{ hostvars[groups['redisservers'][0]]['ansible_' + iface].ipv4.address }}"`
put it gives me the error: 
```
TASK [common : Copying environment variable file] ******************************
[DEPRECATION WARNING]: evaluating development as a bare variable, this 
behaviour will go away and you might need to add |bool to the expression in the
 future. Also see CONDITIONAL_BARE_VARS configuration toggle.. This feature 
will be removed in version 2.12. Deprecation warnings can be disabled by 
setting deprecation_warnings=False in ansible.cfg.
fatal: [default]: FAILED! => {"changed": false, "msg": "AnsibleUndefinedVariable: 'dict object' has no attribute 'redisservers'"}
``` 
during provisioning. 

I have also tried to use a Docker command to get the ip at runtime: `redis_host: "{% raw %}$(sudo docker inspect -f \"{{ .NetworkSettings.Networks.web_network.IPAddress }}\" kleeweb-redis){% endraw %}"` but it does also seem to have issues after provisioning when running the website.

I have spent some time with some possible solutions that failed, but do have to say that it is probably not worth changing this, because this hardcoded IP address is only temporary and will be changed with the dockerized PR into a variable which is not hardcoded.

Denis-Gavrielov commented:
@andronat thanks again for looking over my PR in such detail.

I have now addresses all the issues, made the necessary changes, and made comments answering questions (in some cases also explaining why I would not go with the suggested change).
There is a separate PR (#133) for all the "quick fixes" before the Circle CI update. #133 obviously fails in Circle CI, because no Circle CI updates are done. 

With this PR (#132) testing works in Circle CI as well as in dev.

Denis-Gavrielov commented:
@andronat I have now removed the copied klee-web-environment.sh file as discussed.

andronat commented:
JFYI you can rebase this now.

Denis-Gavrielov commented:
@andronat this branch is now rebased.

ccadar commented:
Great to see this in, thanks again!
I have now changed the settings to require Circle CI to pass before being able to merge a PR.  Let me know if this doesn't work as expected.
